Packs/CortexAttackSurfaceManagement/Scripts/RankServiceOwners/RankServiceOwners.py

```diff
@@ -202,16 +299,300 @@ def _get_k(
+            # abort early
+            return {
+                f"{canonical_first_initial}{''.join([m[0] for m in middle_names])}{last_initial}"
+            }
+        else:
+            for hypothesized_letter in string.ascii_lowercase:
+                result_set.add(
+                    f"{canonical_first_initial}{hypothesized_letter}{last_initial}"
+                )
+
+    return result_set
+
+
+def get_name_similarity_index(
+    personal_monikers: Iterable[str],
+    constant_name: str,
+) -> float:
+    """
+    Returns an index into name similarity between `personal_monikers` and `constant_name`.
+
+    Returns >=1 if there is a blatant match.
+    Returns 0 if there is no match at all.
+    Returns 0 to 1 if there is a potential match.
+    """
+    total_indicators = 0.0
+
+    all_monikers: set[str] = map(  # type: ignore
+        str.lower,
+        generate_all_spaceless_monikers(personal_monikers),
+    )
+    all_monikers = {m for m in all_monikers if len(m) > 1}
+    all_names = split_phrase(constant_name.lower())
+    all_names = {n for n in all_names if len(n) > 1}
+
+    for moniker in all_monikers:
+        if moniker in all_names:
+            demisto.info(f"Name similarity match: {constant_name} and {moniker}")
+            total_indicators += 1
+        else:
+            for n in all_names:
+                if moniker in n:
+                    demisto.info(f"Name substring match: {constant_name} and {moniker}")
+                    total_indicators += 0.01
+
+    # check for a hypothesized-middle-initial match
+    # we want this because gmj-test --> George Jones (gejones)
+    hypothesized_initials = get_possible_3initials(personal_monikers) - all_monikers
+    for hypothesized_initial in hypothesized_initials:
+        if hypothesized_initial in all_names:
+            demisto.info(f"Hypothesized initial match: {constant_name} and {moniker}")
+            total_indicators += 0.1
+
+    return total_indicators
+
+
+class OwnerFeaturizationPipeline():
+    def __init__(self, sources: list | None = None):
+        """
+        Initialize a featurization pipeline.
+
+        As of May 2023, we use `sources` to parse out which sources are included in the
+        RC-provided owner data. The "parse the sources for a parallel list of known
+        sources" approach is quite fragile (it requires us to release a new model every
+        time RC adds a new source if the model wants to use that new source as a new source);
+        we should work with RC to change the data format so that this approach is less
+        fragile.
+        """
+        if sources is None:
+            # Hardcoding is true as of May 2023.
+            # To get the fields available as Sources, within the `content` repo, search the path
+            # `Packs/CortexAttackSurfaceManagement`
+            # for `Name,Email,Source,Timestamp`.
+
+            # FIXME (plt 2023.06): work with RC to retain Source System directly from the `owners` metadata
```

Comment:
@kball-pa let's try to resolve this FIXME before merge.

ideally it looks like we'd have RC add a field upstream that is specifically for the "source system" concept. can you look into whether this is feasible on a reasonable timeframe?

Reply:
see https://github.com/demisto/content/pull/28375#discussion_r1276594560